diff --git a/include/boost/thread/win32/basic_timed_mutex.hpp b/include/boost/thread/win32/basic_timed_mutex.hpp
index f0905e66..800c1769 100644
--- a/include/boost/thread/win32/basic_timed_mutex.hpp
+++ b/include/boost/thread/win32/basic_timed_mutex.hpp
@@ -82,7 +82,7 @@ namespace boost
 
                     do
                     {
-                        if(winapi::WaitForSingleObjectEx(sem,::boost::detail::win32::infinite,0)==0)
+                        if(winapi::WaitForSingleObjectEx(sem,::boost::detail::win32::infinite,1)==0)
                         {
                             clear_waiting_and_try_lock(old_count);
                         }
@@ -182,7 +182,7 @@ namespace boost
                         {
                             d = (std::min)(d, max);
                         }
-                        if(winapi::WaitForSingleObjectEx(sem,getMs(d),0)==0)
+                        if(winapi::WaitForSingleObjectEx(sem,getMs(d),1)==0)
                         {
                             clear_waiting_and_try_lock(old_count);
                         }
diff --git a/include/boost/thread/win32/thread_primitives.hpp b/include/boost/thread/win32/thread_primitives.hpp
index 391d7375..6a2714be 100644
--- a/include/boost/thread/win32/thread_primitives.hpp
+++ b/include/boost/thread/win32/thread_primitives.hpp
@@ -182,7 +182,7 @@ namespace boost
                 else
                 {
 #if BOOST_PLAT_WINDOWS_RUNTIME
-                    ::boost::winapi::WaitForSingleObjectEx(::boost::winapi::GetCurrentThread(), milliseconds, 0);
+                    ::boost::winapi::WaitForSingleObjectEx(::boost::winapi::GetCurrentThread(), milliseconds, 1);
 #else
                     ::boost::winapi::Sleep(milliseconds);
 #endif
diff --git a/src/win32/thread.cpp b/src/win32/thread.cpp
index 000ab877..763f6f61 100644
--- a/src/win32/thread.cpp
+++ b/src/win32/thread.cpp
@@ -341,7 +341,7 @@ namespace boost
         return start_thread_noexcept();
 #else
       uintptr_t const new_thread=_beginthreadex(0,static_cast<unsigned int>(attr.get_stack_size()),&thread_start_function,thread_info.get(),
-                                                CREATE_SUSPENDED | STACK_SIZE_PARAM_IS_A_RESERVATION, &thread_info->id);
+                                                CREATE_SUSPENDED, &thread_info->id);
       if(!new_thread)
       {
         return false;
@@ -496,7 +496,7 @@ namespace boost
     bool thread::interruption_requested() const BOOST_NOEXCEPT
     {
         detail::thread_data_ptr local_thread_info=(get_thread_info)();
-        return local_thread_info.get() && (winapi::WaitForSingleObjectEx(local_thread_info->interruption_handle,0,0)==0);
+        return local_thread_info.get() && (winapi::WaitForSingleObjectEx(local_thread_info->interruption_handle,0,TRUE)==0);
     }
 
 #endif
@@ -681,7 +681,7 @@ namespace boost
             {
                 if(handle_count)
                 {
-                    unsigned long const notified_index=winapi::WaitForMultipleObjectsEx(handle_count,handles,false,static_cast<DWORD>(time_left_msec), 0);
+                    unsigned long const notified_index=winapi::WaitForMultipleObjectsEx(handle_count,handles,false,static_cast<DWORD>(time_left_msec), TRUE);
                     if(notified_index<handle_count)
                     {
                         if(notified_index==wait_handle_index)
@@ -779,7 +779,7 @@ namespace boost
             {
                 if(handle_count)
                 {
-                    unsigned long const notified_index=winapi::WaitForMultipleObjectsEx(handle_count,handles,false,static_cast<DWORD>(time_left_msec), 0);
+                    unsigned long const notified_index=winapi::WaitForMultipleObjectsEx(handle_count,handles,false,static_cast<DWORD>(time_left_msec), TRUE);
                     if(notified_index<handle_count)
                     {
                         if(notified_index==wait_handle_index)
@@ -840,7 +840,7 @@ namespace boost
 
         bool interruption_requested() BOOST_NOEXCEPT
         {
-            return detail::get_current_thread_data() && (winapi::WaitForSingleObjectEx(detail::get_current_thread_data()->interruption_handle,0,0)==0);
+            return detail::get_current_thread_data() && (winapi::WaitForSingleObjectEx(detail::get_current_thread_data()->interruption_handle,0,TRUE)==0);
         }
 #endif
 
