diff --git a/include/boost/interprocess/sync/interprocess_semaphore.hpp b/include/boost/interprocess/sync/interprocess_semaphore.hpp
index abb74e8..f4a336f 100644
--- a/include/boost/interprocess/sync/interprocess_semaphore.hpp
+++ b/include/boost/interprocess/sync/interprocess_semaphore.hpp
@@ -93,7 +93,7 @@ class interprocess_semaphore
    bool timed_wait(const TimePoint &abs_time);
 
    //!Returns the interprocess_semaphore count
-//   int get_count() const;
+   int get_count();
    #if !defined(BOOST_INTERPROCESS_DOXYGEN_INVOKED)
    private:
    #if defined(BOOST_INTERPROCESS_SEMAPHORE_USE_POSIX)
@@ -135,6 +135,9 @@ inline bool interprocess_semaphore::timed_wait(const TimePoint &abs_time)
 inline void interprocess_semaphore::post()
 { m_sem.post(); }
 
+inline int interprocess_semaphore::get_count()
+{ return m_sem.get_count(); }
+
 }  //namespace interprocess {
 }  //namespace boost {
 
diff --git a/include/boost/interprocess/sync/posix/semaphore.hpp b/include/boost/interprocess/sync/posix/semaphore.hpp
index bf64d43..5688070 100644
--- a/include/boost/interprocess/sync/posix/semaphore.hpp
+++ b/include/boost/interprocess/sync/posix/semaphore.hpp
@@ -50,6 +50,9 @@ class posix_semaphore
    bool try_wait()
    {  return semaphore_try_wait(&m_sem); }
 
+   int get_count()
+   { return semaphore_get_count(&m_sem); }
+
    template<class TimePoint>
    bool timed_wait(const TimePoint &abs_time)
    {  return semaphore_timed_wait(&m_sem, abs_time); }
diff --git a/include/boost/interprocess/sync/posix/semaphore_wrapper.hpp b/include/boost/interprocess/sync/posix/semaphore_wrapper.hpp
index c77397a..eb0f62c 100644
--- a/include/boost/interprocess/sync/posix/semaphore_wrapper.hpp
+++ b/include/boost/interprocess/sync/posix/semaphore_wrapper.hpp
@@ -143,6 +143,7 @@ inline bool semaphore_unlink(const char *semname)
 inline void semaphore_init(sem_t *handle, unsigned int initialCount)
 {
    int ret = sem_init(handle, 1, initialCount);
+
    //According to SUSV3 version 2003 edition, the return value of a successful
    //sem_init call is not defined, but -1 is returned on failure.
    //In the future, a successful call might be required to return 0.
@@ -192,6 +193,16 @@ inline bool semaphore_try_wait(sem_t *handle)
    throw interprocess_exception(err);
 }
 
+inline int semaphore_get_count(sem_t *handle)
+{
+   int value;
+   int res = sem_getvalue(handle, &value);
+   if (res == 0)
+      return value;
+   error_info err = system_error_code();
+   throw interprocess_exception(err);
+}
+
 #ifndef BOOST_INTERPROCESS_POSIX_TIMEOUTS
 
 struct semaphore_wrapper_try_wrapper
diff --git a/include/boost/interprocess/sync/spin/semaphore.hpp b/include/boost/interprocess/sync/spin/semaphore.hpp
index cd3c866..f692f13 100644
--- a/include/boost/interprocess/sync/spin/semaphore.hpp
+++ b/include/boost/interprocess/sync/spin/semaphore.hpp
@@ -12,83 +12,76 @@
 #define BOOST_INTERPROCESS_DETAIL_SPIN_SEMAPHORE_HPP
 
 #ifndef BOOST_CONFIG_HPP
-#  include <boost/config.hpp>
+#include <boost/config.hpp>
 #endif
 #
 #if defined(BOOST_HAS_PRAGMA_ONCE)
-#  pragma once
+#pragma once
 #endif
 
-#include <boost/interprocess/detail/config_begin.hpp>
-#include <boost/interprocess/detail/workaround.hpp>
+#include <boost/cstdint.hpp>
 #include <boost/interprocess/detail/atomic.hpp>
+#include <boost/interprocess/detail/config_begin.hpp>
 #include <boost/interprocess/detail/os_thread_functions.hpp>
+#include <boost/interprocess/detail/workaround.hpp>
 #include <boost/interprocess/sync/detail/common_algorithms.hpp>
 #include <boost/interprocess/sync/detail/locks.hpp>
-#include <boost/cstdint.hpp>
 
 namespace boost {
 namespace interprocess {
 namespace ipcdetail {
 
-class spin_semaphore
-{
-   spin_semaphore(const spin_semaphore &);
-   spin_semaphore &operator=(const spin_semaphore &);
+class spin_semaphore {
+  spin_semaphore(const spin_semaphore &);
+  spin_semaphore &operator=(const spin_semaphore &);
+
+public:
+  spin_semaphore(unsigned int initialCount);
+  ~spin_semaphore();
 
-   public:
-   spin_semaphore(unsigned int initialCount);
-   ~spin_semaphore();
+  void post();
+  void wait();
+  bool try_wait();
+  template <class TimePoint> bool timed_wait(const TimePoint &abs_time);
 
-   void post();
-   void wait();
-   bool try_wait();
-   template<class TimePoint> bool timed_wait(const TimePoint &abs_time);
+  int get_count();
 
-//   int get_count() const;
-   private:
-   volatile boost::uint32_t m_count;
+private:
+  mutable volatile boost::uint32_t m_count;
 };
 
+inline spin_semaphore::~spin_semaphore() {}
 
-inline spin_semaphore::~spin_semaphore()
-{}
+inline spin_semaphore::spin_semaphore(unsigned int initialCount) {
+  ipcdetail::atomic_write32(&this->m_count, boost::uint32_t(initialCount));
+}
 
-inline spin_semaphore::spin_semaphore(unsigned int initialCount)
-{  ipcdetail::atomic_write32(&this->m_count, boost::uint32_t(initialCount));  }
+inline void spin_semaphore::post() { ipcdetail::atomic_inc32(&m_count); }
 
-inline void spin_semaphore::post()
-{
-   ipcdetail::atomic_inc32(&m_count);
+inline void spin_semaphore::wait() {
+  ipcdetail::lock_to_wait<spin_semaphore> lw(*this);
+  return ipcdetail::try_based_lock(lw);
 }
 
-inline void spin_semaphore::wait()
-{
-   ipcdetail::lock_to_wait<spin_semaphore> lw(*this);
-   return ipcdetail::try_based_lock(lw);
+inline bool spin_semaphore::try_wait() {
+  return ipcdetail::atomic_add_unless32(&m_count, boost::uint32_t(-1),
+                                        boost::uint32_t(0));
 }
 
-inline bool spin_semaphore::try_wait()
-{
-   return ipcdetail::atomic_add_unless32(&m_count, boost::uint32_t(-1), boost::uint32_t(0));
+template <class TimePoint>
+inline bool spin_semaphore::timed_wait(const TimePoint &abs_time) {
+  ipcdetail::lock_to_wait<spin_semaphore> lw(*this);
+  return ipcdetail::try_based_timed_lock(lw, abs_time);
 }
 
-template<class TimePoint>
-inline bool spin_semaphore::timed_wait(const TimePoint &abs_time)
-{
-   ipcdetail::lock_to_wait<spin_semaphore> lw(*this);
-   return ipcdetail::try_based_timed_lock(lw, abs_time);
+inline int spin_semaphore::get_count() {
+  return (int)ipcdetail::atomic_read32(&m_count);
 }
 
-//inline int spin_semaphore::get_count() const
-//{
-   //return (int)ipcdetail::atomic_read32(&m_count);
-//}
-
-}  //namespace ipcdetail {
-}  //namespace interprocess {
-}  //namespace boost {
+} // namespace ipcdetail
+} // namespace interprocess
+} // namespace boost
 
 #include <boost/interprocess/detail/config_end.hpp>
 
-#endif   //BOOST_INTERPROCESS_DETAIL_SPIN_SEMAPHORE_HPP
+#endif // BOOST_INTERPROCESS_DETAIL_SPIN_SEMAPHORE_HPP
