diff --git a/include/Zycore/Defines.h b/include/Zycore/Defines.h
index 847700b..7ee81ee 100644
--- a/include/Zycore/Defines.h
+++ b/include/Zycore/Defines.h
@@ -64,10 +64,6 @@
 #if defined(__clang__)
 #   define ZYAN_CLANG
 #   define ZYAN_GNUC
-#   if defined(_MSC_VER)
-#       define ZYAN_CLANG_CL
-#       define ZYAN_MSVC
-#   endif
 #elif defined(__ICC) || defined(__INTEL_COMPILER)
 #   define ZYAN_ICC
 #elif defined(__GNUC__) || defined(__GNUG__)
@@ -194,7 +190,7 @@
 /* Generic DLL import/export helpers                                                              */
 /* ============================================================================================== */
 
-#if defined(ZYAN_MSVC) || (defined(ZYAN_WINDOWS) && defined(ZYAN_GNUC))
+#if defined(ZYAN_MSVC)
 #   define ZYAN_DLLEXPORT __declspec(dllexport)
 #   define ZYAN_DLLIMPORT __declspec(dllimport)
 #else
